pawn.cc

```diff
@@ -0,0 +1,112 @@
+#include "pawn.h"
+
+Pawn::Pawn(std::string name, std::vector<std::vector<Piece *>> *board, bool whitePlayer, int xCoord, int yCoord):
+    Piece{name, board,whitePlayer, xCoord, yCoord} {}
+
+bool Pawn::isLegal(Box &targetBox) {
+    // checks if targetBox is occupied and Piece is of opposite colour (in order to capture)
+    if ((((*(this->getBoard()))[targetBox.getX()][targetBox.getY()]) && 
+        ((checkWhitePlayer() && !(((*(this->getBoard()))[targetBox.getX()][targetBox.getY()])->checkWhitePlayer())) ||
+         (!checkWhitePlayer() && (((*(this->getBoard()))[targetBox.getX()][targetBox.getY()])->checkWhitePlayer())) ) )) {
+        return true;
+    }
+    return false;
+}
+
+void Pawn::updateLegalMoves() {
+    std::vector<Box> legalMoves;
+    std::vector<int> moveStates;
+
+    int x = this->getX();
+    int y = this->getY();
+
+    // Caputre
+    if (checkWhitePlayer() && (x - 1 >= 0) && (y - 1 >= 0)) {
+        Box captureMove1(x - 1, y - 1);
+        if (isLegal(captureMove1)) {
+            legalMoves.push_back(captureMove1);
+            moveStates.push_back(1);
+        }
+    }
+    
+    if (checkWhitePlayer() && (x - 1 >= 0) && (y + 1 < 8)) {
+        Box captureMove2(x - 1, y + 1);
+        if (isLegal(captureMove2)) {
+            legalMoves.push_back(captureMove2);
+            moveStates.push_back(1);
+        }
+    }
+
+    if (!checkWhitePlayer() && (x + 1 < 8) && (y - 1 >= 0)) {
+        Box captureMove3(x + 1, y - 1);
+        if (isLegal(captureMove3)) {
+            legalMoves.push_back(captureMove3);
+            moveStates.push_back(1);
+        }
+    }
+
+    if (!checkWhitePlayer() && (x + 1 < 8) && (y + 1 < 8)) {
+        Box captureMove4(x + 1, y + 1);
+        if (isLegal(captureMove4)) {
+            legalMoves.push_back(captureMove4);
+            moveStates.push_back(1);
+        }
+    }
+
+    // First move (2 spaces forward)
+    if (checkWhitePlayer() && (x == 6) && !((*(this->getBoard()))[x - 2][y])) {
+        Box initialMove1(x - 2, y);
+        legalMoves.push_back(initialMove1);
+        moveStates.push_back(0);
+
+        // En passant (Note that we need to delete the newly added move to the opposing pawn 
+        // if it is not used immediately!)
+        if ((y - 1 >= 0) && ((*(this->getBoard()))[x - 2][y - 1])->getName() == "p") {
+            Box opponentPawnMove1(x - 1, y);
+            ((*(this->getBoard()))[x - 2][y - 1])->getLegalMoves()->push_back(opponentPawnMove1);
+            ((*(this->getBoard()))[x - 2][y - 1])->getMoveStates()->push_back(-1);
+        }
```

Comment:
IF you take a look at `piece.cc` you will see that the `updateMoves()` happens right when the player wants to `move` the piece. Hence, it means when it's the other person's turn to move, the `updateMoves()` function will get called and then the en passiant move will get lost b/c a new array of `LegalMoves` is created.

The reason why `updateLegalMoves()` is called first, before the move actually occurs, is because it covers the case when the first move for each player is made. You are not allowed to call any virtual functions (in this case, it's the `updateLegalMoves()` in the constructor of a piece. So the only way to obtain the legalMoves is prior to moving the piece. 

Hence, for this `updateLegalMoves`, in order to not lose the en passiant move, I think we should get the previous moveState array, search through that array to see if there is a `-1` move that represents en passiant, and then just push that `legalMove` back onto the new array.

That way, the en passiant move doesn't get lost and we still able to get around the problem of not being allowed to call the `updateLegalMoves()` in the constructor of a piece.

lmk if that makes sense to you

Reply:
Wondering if this will cause bad behaviour when we first create a piece and call on `updateLegalMoves()` as we are trying to access the "previous" `legalMove` array which is nothing

Reply:
I think in Piece constructor we can have `legalMoveStates` initialized to nullptr
`Piece::Piece(std::string name, std::vector<std::vector<Piece *>> *board, bool whitePlayer, const int xCoord, const int yCoord):
    name{name}, board{board}, whitePlayer{whitePlayer}, xCoord{xCoord}, yCoord{yCoord}, legalMoveStates{nullptr} {};`